muschema/src/array.ts

```diff
@@ -19,20 +28,180 @@ export class MuArray<ValueSchema extends MuSchema<any>>
         };
     }
 
-    public alloc() { return this.identity.slice(); }
+    public alloc () : _MuArrayType<ValueSchema> {
+        return this._pool.pop() || this.identity.slice();
+    }
 
-    public free(x:ValueSchema['identity'][]) { }
+    public free (x:_MuArrayType<ValueSchema>) : void {
+        this._pool.push(x);
+
+        const valueSchema = this.muData;
+        switch (valueSchema.muType) {
+            case 'boolean':
+            case 'float32':
+            case 'float64':
+            case 'int8':
+            case 'int16':
+            case 'int32':
+            case 'string':
+            case 'uint8':
+            case 'uint16':
+            case 'uint32':
+                break;
+            default:
+                for (let i = 0; i < x.length; ++i) {
+                    valueSchema.free(x[i]);
+                }
+        }
+    }
+
+    public clone (x:_MuArrayType<ValueSchema>) : _MuArrayType<ValueSchema> {
+        const result = this._pool.pop() || new Array(x.length);
 
-    public clone(x:ValueSchema['identity'][]) : ValueSchema['identity'][] {
-        const result:ValueSchema['identity'][] = new Array(x.length);
         const schema = this.muData;
-        for (let i = 0; i < x.length; ++i) {
-            result[i] = schema.clone(x[i]);
+        switch (schema.muType) {
+            case 'boolean':
+            case 'float32':
+            case 'float64':
+            case 'int8':
+            case 'int16':
+            case 'int32':
+            case 'string':
+            case 'uint8':
+            case 'uint16':
+            case 'uint32':
+                for (let i = 0; i < x.length; ++i) {
+                    result[i] = x[i];
+                }
+                break;
+            default:
+                for (let i = 0; i < x.length; ++i) {
+                    result[i] = schema.clone(x[i]);
+                }
+        }
+
+        return result;
+    }
+
+    public diffBinary (base:_MuArrayType<ValueSchema>, target:_MuArrayType<ValueSchema>, ws:MuWriteStream) {
+        const targetLength = target.length;
+        const lengthDiff = base.length - target.length;
+        const numDelete = lengthDiff > 0 ? lengthDiff : 0;
+        ws.writeUint32(numDelete);
```

Comment:
what if instead we just write the length of the array instead?

Reply:
yea, you're right. That's more straightforward